pkg/ddl/notifier/publish_testkit_test.go

```diff
@@ -33,7 +35,7 @@ func TestPublishToTableStore(t *testing.T) {
 CREATE TABLE ddl_notifier (
 	ddl_job_id BIGINT,
 	multi_schema_change_seq BIGINT COMMENT '-1 if the schema change does not belong to a multi-schema change DDL. 0 or positive numbers representing the sub-job index of a multi-schema change DDL',
-	schema_change JSON COMMENT 'SchemaChange at rest',
+	schema_change LONGBLOB COMMENT 'SchemaChange at rest',
```

Comment:
What's the reason to change the column type?

Reply:
I tried the `JSON` type and use `row.GetJSON()` and `row.GetBytes()` get the serialized data, but it can't be deserialized. I took a look at other system tables like `tidb_global_task` and `tidb_background_subtask`, they use `LONGBLOB ` type save the meta and `row.GetBytes()` get the serialized data and I did so. Is it my way of operating the `JSON` type is wrong?

Reply:
```diff
diff --git a/pkg/ddl/notifier/publish_testkit_test.go b/pkg/ddl/notifier/publish_testkit_test.go
index 30f0c1170f..d5e01cbabf 100644
--- a/pkg/ddl/notifier/publish_testkit_test.go
+++ b/pkg/ddl/notifier/publish_testkit_test.go
@@ -35,7 +35,7 @@ func TestPublishToTableStore(t *testing.T) {
 CREATE TABLE ddl_notifier (
        ddl_job_id BIGINT,
        multi_schema_change_seq BIGINT COMMENT '-1 if the schema change does not belong to a multi-schema change DDL. 0 or positive numbers representing the sub-job index of a multi-schema change DDL',
-       schema_change LONGBLOB COMMENT 'SchemaChange at rest',
+       schema_change JSON COMMENT 'SchemaChange at rest',
        processed_by_flag BIGINT UNSIGNED DEFAULT 0 COMMENT 'flag to mark which subscriber has processed the event',
        PRIMARY KEY(ddl_job_id, multi_schema_change_seq)
 )
diff --git a/pkg/ddl/notifier/store.go b/pkg/ddl/notifier/store.go
index c652fd0052..f259277e0b 100644
--- a/pkg/ddl/notifier/store.go
+++ b/pkg/ddl/notifier/store.go
@@ -94,7 +94,12 @@ func (t *tableStore) List(ctx context.Context, se *sess.Session, limit int) ([]*
        ret := make([]*schemaChange, 0, len(rows))
        for _, row := range rows {
                event := SchemaChangeEvent{}
-               err = json.Unmarshal(row.GetBytes(2), &event)
+               binaryJSON := row.GetJSON(2)
+               jsonStr, err := binaryJSON.MarshalJSON()
+               if err != nil {
+                       return nil, errors.Trace(err)
+               }
+               err = json.Unmarshal(jsonStr, &event)
                if err != nil {
                        return nil, errors.Trace(err)
                }
```

However I don't have strong opinion about the column type. JSON type can add JSON syntax check in the SQL layer and readable SELECT output, LONGTEXT has readable SELECT output. LONGBLOB has no advantage. Maybe wait other reviewers' opinions.

Reply:
I tried again and found that the `JSON` type in table is not equal to our internal `json` type in code, we need a conversion to use the `JSON` type in table like this:
```
j, err := row.GetJSON(2).MarshalJSON()
err = json.Unmarshal(j, &event)
```
 https://github.com/pingcap/tidb/blob/master/pkg/types/json_binary.go#L151-L180